### PR TITLE
WIP: analyze options saved and used by auto analyze

### PR DESCRIPTION
**Pull request by chrysan**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:

### Check List 

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
Please add a release note, or a 'None' if it is not needed.
```

### CODE REVIEW

File: executor/analyze.go [hunk 1]
```diff
@@ -85,6 +86,11 @@ const (
 
 // Next implements the Executor Next interface.
 func (e *AnalyzeExec) Next(ctx context.Context, req *chunk.Chunk) error {
+	err0 := saveOptsForAutoAnalyze(e)
+	if err0 != nil {
+		logutil.Logger(ctx).Warn("Fail to save options for auto analyze")
```
Comment (winoros) [inline executor/analyze.go:91]: We can also output some warnings to the MySQL client. The session context held in `AnalyzeExec` can do this by calling `AnalyzeExec.ctx.GetSessionVars().StmtCtx.AppendWarning()`

Comment (chrysan) [reply executor/analyze.go:91]: Good point. Done.

File: util/codec/codec.go [hunk 2]
```diff
@@ -898,6 +898,9 @@ func CutOne(b []byte) (data []byte, remain []byte, err error) {
 	if err != nil {
 		return nil, nil, errors.Trace(err)
 	}
+	if l < 0 {
+		return nil, nil, errors.New("Fail to peek the first encoded value")
+	}
```
Comment (chrysan) [inline util/codec/codec.go:903]: @tangenta please review this fix for failed ut related with issues/27434

File: session/bootstrap.go [hunk 3]
```diff
@@ -348,6 +348,14 @@ const (
 		key idx(filter_type),
 		primary key(id)
 	);`
+
+	// CreateAnalyzeOptTable stores analyze options for the usage of auto analyze
+	CreateAnalyzeOptTable = `CREATE TABLE IF NOT EXISTS mysql.analyze_opt(
+		tid bigint(64),
+		opt_key char(255) NOT NULL DEFAULT '',
+		opt_val char(32) NOT NULL DEFAULT '',
+		PRIMARY KEY (tid,opt_key)
```
Comment (xuyifangreeneyes) [inline session/bootstrap.go:357]: Maybe we can use clustered index here.

**Comment by xuyifangreeneyes** [comment]
@chrysan @winoros I have a question that may be not so related to the pr: is there any way to save column options? Column options would be like
```
enum ColumnOption {
    AllColumns,
    PredicateColumns,
    ListedColumns(Vec<int64>)
}
```
IMHO it is kind of hard to be saved with schema and maybe we can use json.

File: util/codec/codec.go [hunk 4]
```diff
@@ -898,6 +898,9 @@ func CutOne(b []byte) (data []byte, remain []byte, err error) {
 	if err != nil {
 		return nil, nil, errors.Trace(err)
 	}
+	if l < 0 {
```
Comment (tangenta) [inline util/codec/codec.go:901]:
```suggestion
	if l < 0 || l > len(b) {
```

**Pull request closed without merging**